NitroxClient/GameLogic/Cyclops.cs

```diff
@@ -53,5 +75,452 @@ public void ActivateShield(string guid)
+        /// it's packaged and sent out to the server if the player is the current owner of the Cyclops. This can trigger sending <see cref="CyclopsDamage"/> packets
+        /// </summary>
+        public void OnTakeDamage(SubRoot subRoot, Optional<DamageInfo> info)
+        {
+            BroadcastDamageChange(subRoot, info);
+        }
+
+        /// <summary>
+        /// Called when the player repairs a <see cref="CyclopsDamagePoint"/>. Right now it's not possible to partially repair because it would be difficult to implement.
+        /// <see cref="CyclopsDamagePoint"/>s are coupled with <see cref="LiveMixin"/>, which is used with just about anything that has health.
+        /// I would need to hook onto <see cref="LiveMixin.AddHealth(float)"/>, or maybe the repair gun event to catch when something repairs a damage point, which I don't 
+        /// believe is worth the effort. A <see cref="CyclopsDamagePoint"/> is already fully repaired in a little over a second. This can trigger sending 
+        /// <see cref="CyclopsDamagePointHealthChanged"/> and <see cref="CyclopsDamage"/> packets
+        /// </summary>
+        public void OnDamagePointHealthChanged(SubRoot subRoot, CyclopsDamagePoint damagePoint, float repairAmount)
+        {
+            string subGuid = GuidHelper.GetGuid(subRoot.gameObject);
+
+            for (int i = 0; i < subRoot.damageManager.damagePoints.Length; i++)
+            {
+                if (subRoot.damageManager.damagePoints[i] == damagePoint)
+                {
+                    CyclopsDamagePointHealthChanged packet = new CyclopsDamagePointHealthChanged(subGuid, i, repairAmount);
+                    packetSender.Send(packet);
+
+                    BroadcastDamageChange(subRoot, Optional<DamageInfo>.Empty());
+
+                    return;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Set the health of a <see cref="CyclopsDamagePoint"/>. This can trigger sending <see cref="CyclopsDamagePointHealthChanged"/>
+        /// and <see cref="CyclopsDamage"/> packets
+        /// </summary>
+        /// <param name="repairAmount">The max health of the point is 1. 999 is passed to trigger a full repair of the <see cref="CyclopsDamagePoint"/></param>
+        public void SetDamagePointHealth(SubRoot subRoot, int damagePointIndex, float repairAmount)
+        {
+            subRoot.damageManager.damagePoints[damagePointIndex].liveMixin.AddHealth(repairAmount);
+        }
+
+        /// <summary>
+        /// This is run during every major state change. If this player is the current owner, a <see cref="CyclopsDamage"/> packet will be sent out
+        /// </summary>
+        private void BroadcastDamageChange(SubRoot subRoot, Optional<DamageInfo> info)
+        {
+            string subGuid = GuidHelper.GetGuid(subRoot.gameObject);
+
+            // If it isn't the owner, we don't want to broadcast
+            if (!simulationOwnershipManager.HasAnyLockType(subGuid))
```

Comment:
Do we want to prevent this damage from happening in the first place?  Seems like at this point it is out of sync.

Reply:
This was one of the things I realized I missed just after committing. Currently, the fires are blocked from being created client side by an ownership check at SubFire_CreateFire_Patch.

Damage points don't get blocked like fires do, meaning there can be a de-sync between damage points until the owner broadcasts a new CyclopsDamage packet. I'll add a patch to CyclopsExternalDamageManager.CreatePoint() to intercept damage point creation.

I was hesitant to hook onto and block SubRoot.OnTakeDamage only because it's a sizeable chunk of code that doesn't do much beyond shake the player's screen. It's possible for one client to get a strike warning while the others do not, but the damage status of the Cyclops should remain unchanged if I hook onto and block the damage point creation event.

I think what I'll do is block the function if they aren't the owner, but have it run the client side sound and screen shake for non-owners. That way it'll feel responsive but also leave the Cyclops unaffected until the owner verifies it was an actual hit. Phantom hits should be a rare occurrence, but if they are too common, it'll be easy enough to remove.